### DISCUSSION TITLE
Caching of constant expressions

### DISCUSSION THREAD

**Discussion by erikfroseth** [duckdb/duckdb#6599] [Ideas]
Hello!

I have a query that look roughly like this:
```
SELECT
  key,
  COUNT(DISTINCT value) FILTER (WHERE timestamp_utc > NOW() - INTERVAL 1 DAY) AS last_day,
  COUNT(DISTINCT value) FILTER (WHERE timestamp_utc > NOW() - INTERVAL 2 DAY) AS last_2_days,
  COUNT(DISTINCT value) FILTER (WHERE timestamp_utc > NOW() - INTERVAL 3 DAY) AS last_3_days
FROM table
GROUP BY key;
```
With around 2.5 million rows in my table, it takes around 1 second to execute (which I'm more than happy with). But if I rewrite the query to use constant string values for the interval expressions (i.e., `2023-03-06T21:00:00` instead of `NOW() - INTERVAL 1 DAY`), the execution time drops down to 0.4 seconds! So it seems like there is an optimization opportunity here: cache and reuse constant expressions. If I understand the DuckDB manual correctly, NOW() should return the start of the current transaction, so its value should remain constant througout the query.

**Comment by Mytherin** [comment]
`NOW` in general cannot be converted to a constant in the regular optimizer pipeline because, while the value is indeed constant for a given execution of a query, queries can be run over multiple transactions when used using e.g. prepared statements. While this is technically possible the optimization is not very clean from a system perspective because it would require a separate optimization step/pipeline that runs outside of the normal optimizer routine.

In general I am quite surprised the expression execution has such an effect, though. Compared to a distinct aggregate the execution of the interval subtraction should be extremely cheap. 

I have investigated a bit, and found that the cast of `timestamp -> timestamptz` is extremely expensive, which happens if `timestamp_utc` is of type `TIMESTAMP` (instead of `TIMESTAMPTZ`). Perhaps that is what you are running into?

The two queries have the same cost when `timestamp_utc` is of type `timestamptz`:

```sql
CREATE TABLE table_with_tz AS
SELECT row_number() OVER () % 5 AS KEY,
              row_number() OVER () AS value,
              timestamp_utc::timestamptz AS timestamp_utc
FROM generate_series(NOW() - interval '1' MONTH, NOW(), interval '10 milliseconds') t(timestamp_utc);
-- roughly 241 million rows

SELECT
  key,
  COUNT(DISTINCT value) FILTER (WHERE timestamp_utc > (NOW() - INTERVAL 1 DAY)) AS last_day,
  COUNT(DISTINCT value) FILTER (WHERE timestamp_utc > (NOW() - INTERVAL 2 DAY)) AS last_2_days,
  COUNT(DISTINCT value) FILTER (WHERE timestamp_utc > (NOW() - INTERVAL 3 DAY)) AS last_3_days
FROM table_with_tz
GROUP BY key;
-- 1.25s

SELECT
  key,
  COUNT(DISTINCT value) FILTER (WHERE timestamp_utc > (timestamptz '2023-03-06 23:04:16.212+01' - INTERVAL 1 DAY)) AS last_day,
  COUNT(DISTINCT value) FILTER (WHERE timestamp_utc > (timestamptz '2023-03-06 23:04:16.212+01' - INTERVAL 2 DAY)) AS last_2_days,
  COUNT(DISTINCT value) FILTER (WHERE timestamp_utc > (timestamptz '2023-03-06 23:04:16.212+01' - INTERVAL 3 DAY)) AS last_3_days
FROM table_with_tz
GROUP BY key;
-- 1.22s
```

But the former query is significantly more expensive if `timestamp_utc` is of type `TIMESTAMP`:

```sql
CREATE TABLE table_without_tz AS
SELECT row_number() OVER () % 5 AS KEY,
       row_number() OVER () AS value,
       timestamp_utc::timestamp AS timestamp_utc
FROM generate_series(NOW() - interval '1' MONTH, NOW(), interval '10 milliseconds') t(timestamp_utc);

SELECT
  key,
  COUNT(DISTINCT value) FILTER (WHERE timestamp_utc > (NOW() - INTERVAL 1 DAY)) AS last_day,
  COUNT(DISTINCT value) FILTER (WHERE timestamp_utc > (NOW() - INTERVAL 2 DAY)) AS last_2_days,
  COUNT(DISTINCT value) FILTER (WHERE timestamp_utc > (NOW() - INTERVAL 3 DAY)) AS last_3_days
FROM table_without_tz
GROUP BY key;
-- 4.9s

SELECT
  key,
  COUNT(DISTINCT value) FILTER (WHERE timestamp_utc > (timestamptz '2023-03-06 23:04:16.212+01' - INTERVAL 1 DAY)) AS last_day,
  COUNT(DISTINCT value) FILTER (WHERE timestamp_utc > (timestamptz '2023-03-06 23:04:16.212+01' - INTERVAL 2 DAY)) AS last_2_days,
  COUNT(DISTINCT value) FILTER (WHERE timestamp_utc > (timestamptz '2023-03-06 23:04:16.212+01' - INTERVAL 3 DAY)) AS last_3_days
FROM table_without_tz
GROUP BY key;
-- 1.22s
```

This query also seems very expensive:

```sql
SELECT min(timestamp_utc::timestamptz) FROM table_without_tz;
-- 3.83s
```

@hawkfish could you perhaps have a look to see if this can be optimized?

**Reply by hawkfish** [reply]
Think this is a cast that is a copy instead of reinterpreting the data.

**Reply by hawkfish** [reply]
Converted to #6604

**Reply by erikfroseth** [reply]
> I have investigated a bit, and found that the cast of `timestamp -> timestamptz` is extremely expensive, which happens if `timestamp_utc` is of type `TIMESTAMP` (instead of `TIMESTAMPTZ`). Perhaps that is what you are running into?
> 

This is definitely what I'm seeing. Changing my column to `timestamptz` gives me the speedup I've been seeing. And thank you for the explanation regarding caching of expressions, especially `NOW()`; I did not think about the complexity around multi-transaction statements.
